### PR TITLE
[DLLM]  vectorized joint/low-confidence decoding and skip redundant attn init

### PR DESCRIPTION
**Pull request by wenxuewuhd**
## Motivation

The Joint Threshold and Low Confidence DLLM parallel decoding paths relied heavily on Python loops over the batch dimension, with frequent per-sample synchronization (e.g. .item()), which limits GPU utilization and throughput.
In addition, each forward in the main decode loop performed a full attention backend initialization, which is redundant across iterations.

This PR speeds up parallel decoding by replacing those hot paths with batched tensor operations (vectorized decoding) while preserving semantics aligned with the original per-sample logic, and by skipping redundant attention backend initialization between steps to cut Python and setup overhead.

## Motivation
The **Joint Threshold** and **Low Confidence** DLLM parallel decoding paths relied heavily on **Python loops over the batch dimension**, with frequent **per-sample synchronization** (e.g. `.item()`), which limits GPU utilization and throughput.
In addition, each `forward` in the main decode loop performed a **full attention backend initialization**, which is **redundant** across iterations.
This PR speeds up parallel decoding by replacing those hot paths with **batched tensor operations (vectorized decoding)** while preserving semantics aligned with the original per-sample logic, and by **skipping redundant attention backend initialization** between steps to cut Python and setup overhead.
### 1. `joint_threshold.py` (Joint Threshold)
- Added `joint_threshold_update_step_vectorized`: implements one decoding step (penalty, argmax / log-softmax confidence, M2T / T2T, post-edit step count, `finished`, etc.) as **tensor ops over `[B, block]`** instead of `for i in range(batch_size)`.
- Introduced **`vectorized_decoding`** in algorithm config; when enabled, builds `prompt_masks` / `start_list` via a **2D view** in one shot instead of per-block Python loops.
- After the first `forward` in the main loop, passes **`skip_attn_backend_init`** on later iterations to avoid re-initializing the attention backend every step; performs a final `forward` when needed to refresh KV / logits.
- A follow-up commit **unifies** the “extra forward” path for vectorized vs non-vectorized code (scalar `any_changed_in_last_step`, single branch) to reduce duplication.
### 2. `low_confidence.py` (Low Confidence)
- Added `parallel_decoding_update_input_ids_vectorized`: **vectorized** threshold-based updates on mask positions using argmax confidence, including the “force at least one pick per row” behavior to match the old implementation, with fewer `batch_id` loops and sync points.
- Same **`vectorized_decoding`** flag; when on, **`start_list`** is derived like Joint Threshold via `prompt_masks.sum(dim=1)`. The main loop uses **`skip_attn_backend_init`**, and the closing `forward` passes the same flag.
### 3. Defaults
- Set the default of `algorithm_config["vectorized_decoding"]` to **`True`** for both Joint Threshold and Low Confidence so the optimized path is **on by default**; set it to `False` explicitly if you need the legacy behavior.

## Benchmarking and Profiling

LLaDA2.1-mini "JointThreshold" bs1-tp1 on 910b with gsm8k zero shot test first 200 (similar acceleration can be obtained on other hardware）

### before
Accuracy: 0.840
Invalid: 0.000
Latency: 161.623 s
Output throughput: 274.918 token/s
metrics={'accuracy': np.float64(0.84), 'invalid': np.float64(0.0), 'latency': 161.6227790140547, 'output_throughput': 274.91793094422735}
image
### after
Accuracy: 0.840
Invalid: 0.000
Latency: 148.679 s
Output throughput: 298.852 token/s
metrics={'accuracy': np.float64(0.84), 'invalid': np.float64(0.0), 'latency': 148.67891802708618, 'output_throughput': 298.85205373841393}
image

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #30432744591](https://github.com/sgl-project/sglang/actions/runs/30432744591)
Latest PR Test (Extra): :x: [Run #30432744305](https://github.com/sgl-project/sglang/actions/runs/30432744305)

### CODE REVIEW

File: python/sglang/srt/dllm/algorithm/low_confidence.py [hunk 1]
```diff
@@ -39,23 +147,46 @@ def run(
...
             logits_output, can_run_cuda_graph = out.logits_output, out.can_run_graph
             assert batch_size == forward_batch.input_ids.shape[0] // self.block_size
+            if self.vectorized_decoding:
+                parallel_decoding_update_input_ids_vectorized(
+                    input_ids_1d=forward_batch.input_ids,
+                    full_logits_2d=logits_output.full_logits,
+                    mask_id=self.mask_id,
+                    block_size=self.block_size,
+                    threshold=self.threshold,
+                    finished=None,  # 如果你有 finished mask，就传进来
```
[the remark this answers is not in the document]
Comment (wenxuewuhd) [reply python/sglang/srt/dllm/algorithm/low_confidence.py]: fixed already

File: python/sglang/srt/dllm/algorithm/joint_threshold.py [hunk 2]
```diff
@@ -9,6 +11,96 @@
...
+
+    # eligible rows (match original semantics)
+    eligible = active & (~(no_mask_active & exceeded))
+
+    # ---------- M2T ----------
+    log_thr = math.log(threshold)
+    neg_inf = torch.full_like(logp, float("-inf"))
+    conf_m2t = torch.where(mask_pos, logp, neg_inf)
+
+    m2t = (conf_m2t > log_thr) & (eligible & has_mask).view(B, 1)
```
[the remark this answers is not in the document]
Comment (wenxuewuhd) [reply python/sglang/srt/dllm/algorithm/joint_threshold.py]: fixed already

**Comment by ping1jing2** [comment]
@wenxuewuhd could you please resolve the conflicts

**Review by ping1jing2 — APPROVED** [review]
